docs/src/main/sphinx/language/types.md

```diff
@@ -159,12 +159,21 @@ Single quotes in string literals can be escaped by using another single quote:
 
 ### `CHAR`
 
-Fixed length character data. A `CHAR` type without length specified has a default length of 1.
-A `CHAR(x)` value always has `x` characters. For example, casting `dog` to `CHAR(7)`
-adds 4 implicit trailing spaces. Leading and trailing spaces are included in comparisons of
-`CHAR` values. As a result, two character values with different lengths (`CHAR(x)` and
-`CHAR(y)` where `x != y`) will never be equal. As with `VARCHAR`, a single quote in a `CHAR` 
+Fixed length character data. A `CHAR` type without length specified has a
+default length of 1. A `CHAR(x)` value always has `x` characters. For example,
+casting `dog` to `CHAR(7)` adds 4 implicit trailing spaces. Leading and trailing
+spaces are included in comparisons of `CHAR` values. As a result, two character
```

Comment:
The following sentence is misleading and irrelevant to an end user. It's a technicality that they will never observe. In fact, it would be more accurate to say that values of `char` types of different length are not directly comparable to each other.

I would just remove that sentence.

Reply:
I am not sure what to remove and leave behind.

Remove this:

As a result, two character values with different lengths (`CHAR(x)` and `CHAR(y)` where `x != y`) are never
equal, but comparison of such values implicitly converts the types to the same
length and pads with spaces so that the following query returns `true`:

and the example code snippet as well?

Reply:
I would reframe the part that starts with "Leading and trailing spaces are included ...". That is not very useful, as it's also how VARCHARs work.

This has nothing to do with comparisons, but with the semantics of casting from VARCHAR or CHAR with smaller length to CHAR with longer length:

```
CAST(CHAR 'abc' AS char(5)) => CHAR 'abc  '
CAST(VARCHAR 'abc' AS char(5)) => CHAR 'abc  '
```

Reply:
Hm .. I think I am confused

Why is the result of this query `true` then?

```
SELECT cast('example' AS char(20)) = cast('example    ' AS char(25));
```

isnt that comparing a char(20) with a char(25) and it should therefore return `false`

Reply:
No, it’s comparing two char(25). There’s an implicit cast that turns the char(20) into a char(25). 

You can see what’s happing more clearly if you EXPLAIN that query.

Reply:
I must be missing something as I cant get EXPLAIN to reveal more useful info. I thought you said that comparison does not do implicit casting. So what is causing this implicit cast.. honestly it feels a bit like this is a bug or at least unexpected behavior and hence I think it will be important to figure out a good documentation for it.. and it might be wider than just char and varchar stuff

Reply:
See

```
SELECT typeof(cast('example' AS char(20))), typeof(cast('example    ' AS char(25))), cast('example' AS char(20)) = cast('example    ' AS char(25));
  _col0   |  _col1   | _col2
----------+----------+-------
 char(20) | char(25) | true
```

The first two columns are confirming that the char length is different .. so they are different. But then the comparison says they are identical.

Reply:
So then the sentence

```
As a result, two character values with different lengths (`CHAR(x)` and
`CHAR(y)` where `x != y`) will never be equal.
```

that is in the docs now .. is kinda misleading ..

Reply:
> I thought you said that comparison does not do implicit casting.

Comparisons require operands to be of the same type. If they are not, the engine will insert implicit casts if the types involved allow so. That's why you can compare, say, a double and a bigint, a char and a varchar, etc. It also applies to any function call that takes a type X and receives an argument of type Y that has an implicit coercion defined between Y -> Y.

Reply:
> The first two columns are confirming that the char length is different .. so they are different. But then the comparison says they are identical.

See the EXPLAIN plan I posted above.

Reply:
> that is in the docs now .. is kinda misleading ..

Yes, while it's technically correct (in the same sense that BIGINT '1' is not equal to DOUBLE '1.0' because the types are not even comparable to each other), it's useless to a user and misleading, because the engine will introduce implicit casts to make that comparison possible.